### DISCUSSION TITLE
Can the FakeTimeProvider (or other mock implementation) be used with Quartz 4?

### DISCUSSION THREAD

**Discussion by mikeohlas** [quartznet/quartznet#2540] [Q&A] [answered]
I am trying to do tests that depend on time elapsed and amount of times jobs been executed. The idea is to replace `TimeProvider.System` with `FakeTimeProvider` in order to control elapsed time manually, eliminating "flakiness" from tests when using delays. 

Its seems that registering FakeTimeProvider: `services.AddSingleton<TimeProvider>(_ => fakeTimeProvider);` does effect the Quartz but no matter the advancement of time with calls such as:
`fakeTimeProvider.Advance(TimeSpan.FromMinutes(1));` the IJob that was registered never fires.

I checked if the FakeTimeProvider is being called with `fakeTimeProvider.AutoAdvanceAmount = TimeSpan.FromSeconds(1);` and it shows that its being called ~5 times.

If I remove the `services.AddSingleton<TimeProvider>(_ => fakeTimeProvider);` line, IJob fires as expected.

my quartz registration code:
```
services.AddQuartz(q =>
                        {
                            var jobKey = new JobKey("Run the domain event check job");
                            q.AddJob<TimedEventChecker>(opts =>
                            {
                                opts.WithIdentity(jobKey);
                            });

                            q.AddTrigger(opts =>
                            {
                                opts.ForJob(jobKey)
                                .WithIdentity(jobKey.Name + "trigger")
                                .StartNow()
                                .WithSimpleSchedule(x =>
                                {
                                    x.WithInterval(TimeSpan.FromMilliseconds(eventCheckerMilis))
                                    .RepeatForever();

                                });
                            });
                            Log.Logger.Information("eventCheckerTimeout delay set to {exuteDelay} mili-seconds", eventCheckerMilis);
                        });

                    services.AddQuartzHostedService(q =>
                    {
                        q.WaitForJobsToComplete = true;
                        q.AwaitApplicationStarted = true;
                    });
```

Is this something Quartz is/will be intended to do or I am doing/understanding something wrong? Thank you

Version info:
Quartz: 4.0.0-preview-20240827-0152
.NET SDK: 8.0.401
OS: Ubuntu 22.04.4 LTS

**Comment by lahma** [comment, marked as answer]
I think it might make it a bit hard to see how system is working if you use a global TimeProvider and expect to see all components interacting. There's so many moving parts and checks. Testing more specific components with TimeProvider would probably be far easier.
